rsi_service/src/lib.rs

```diff
@@ -1,8 +1,251 @@
+    rsi::{PriceData, RsiData, rsi_server},
+    monitor::{PingRequest, PingResponse, monitor_server},
+};
+
+pub fn gen_prost_ts() -> ::prost_types::Timestamp {
+    let ct = SystemTime::now().duration_since(SystemTime::UNIX_EPOCH).unwrap();
+    let mut cs = ct.as_secs();
+    let cn = ct.as_nanos() - ((ct.as_secs() * 1000000000) as u128);
+
+    // This ping function will serve inaccurate timestamps starting from Fri, 11 Apr 2262 23:47:16 UTC
+    if cs > i64::MAX as u64 {
+        cs = i64::MAX as u64;
+    }
+
+    ::prost_types::Timestamp {
+        seconds: cs as i64,
+        nanos: cn as i32
+    }
+}
+
+fn gen_debug_data(uuid: Option<String>) -> DebugData {
+    let uuid = match uuid {
+        Some(s) => s,
+        None => Uuid::new_v4().as_simple().to_string(),
+    };
+
+    DebugData { 
+        ts: Some(gen_prost_ts()),
+        uuid,
+    }
+}
+
+impl DebugData {
+    fn get_uuid(&self) -> String {
+        let DebugData { ts: _, uuid } = self;
+        return uuid.clone();
+    }
+}
+
+fn calc_rsi(pd: Vec<Decimal>) -> String {
+    let mut up: Vec<Decimal> = Vec::with_capacity(14);
+    let mut down: Vec<Decimal> = Vec::with_capacity(14);
+
+    let _ = pd.windows(2).inspect(|arr| {
+        if arr[1] > arr[0] {
+            up.push(arr[1] - arr[0]);
+            down.push(Decimal::ZERO);
+        } else if arr[1] < arr[0] {
+            up.push(Decimal::ZERO);
+            down.push(arr[0] - arr[1]);
+        } else {
+            up.push(Decimal::ZERO);
+            down.push(Decimal::ZERO);
+        }
+    }).collect::<Vec<_>>();
+
+    let rs = calc_smma(up) / calc_smma(down);
+
+    let rsi = Decimal::from(100) - (Decimal::from(100)/(Decimal::from(1)+rs));
+
+    return rsi.to_string();
+}
+
+fn calc_smma(pd: Vec<Decimal>) -> Decimal {
+    pd.iter().enumerate().fold(dec!(0), |accum, e| {
+        (accum * Decimal::from(e.0) + e.1)/(Decimal::from(e.0 + 1))
+    })
+}
+
+#[derive(Debug, Default)]
+pub struct MyMonitor {}
+
+#[tonic::async_trait]
+impl monitor_server::Monitor for MyMonitor {
+    async fn ping(&self, request: Request<PingRequest>) -> Result<Response<PingResponse>, Status> {
+        println!("Got a request: {:?}", request);
+
+        let reply = PingResponse {
+            response_time: Some(gen_prost_ts()),
+            version: String::from("0.0.1")
+        };
+
+        Ok(Response::new(reply))
+    }
+}
+
+#[derive(Debug, Default)]
+pub struct MyRsi {}
+
+#[tonic::async_trait]
+impl rsi_server::Rsi for MyRsi {
+    async fn get_rsi(&self, request: Request<PriceData>) -> Result<Response<RsiData>, Status> {
+        let PriceData { pd, debug } = request.into_inner();
+
+        if pd.len() < 3 || pd.len() > 80 {
+            return Err(Status::new(tonic::Code::InvalidArgument, "too many or too few price values"));
+        }
+
+        let pd: Vec<Decimal> = pd.iter().map(|x| { Decimal::from_str(x).unwrap() }).collect();
+
+
+        let rsival = calc_rsi(pd);
+
+        if let Some(debug) = debug {
```

Comment:
why is this check necessary -- what is the scenario in which you won't have any "debug" data?

Reply:
under normal circumstances you want all replies (successful or otherwise) to throw back the `uuid` specified by the client so the latter can correlate the response

Reply:
The autogenerated code puts the DebugData struct inside of an Option enum. Therefore, I cannot expect a value based on the code.